### DISCUSSION TITLE
Implementing or using DSSIM within libvips

### DISCUSSION THREAD

**Discussion by shun-recall** [libvips/libvips#4221] [General]
I'm trying to optimise a collection of PNG images to JPEG and would like to choose the optimal quality to encode without degrading the visual quality.

I have found that some algorithms use DSSIM to compare the visual degradation of the encoded images.

Is something like DSSIM available or is it possible to do this with libvips?

Thanks.

**Comment by jcupitt** [comment]
Hi @shun-recall,

No, libvips doesn't have a ready-made SSIM operator, but it'd be easy in something like pyvips, just type in the equation (more or less). 

If you don't want to write and test SSIM yourself (understandable!), you could use numpy, I'm sure there are dozens of implementations to chose from. With skimage (first hit on google) you could do something like (untested):

```python
from skimage.metrics import structural_similarity as ssim

a = pyvips.Image.new_from_file("xxxxx.yyy")
b = pyvips.Image.new_from_file("zzzzz.yyy")
s = pyvips.Image.new_from_array(ssim(a.numpy(), b.numpy()))
```

I'm a little wary of these metrics personally. The codec designers will have spent years (with SSIM, focus groups, expert panels, etc.) tuning everything and then giving you a single Q lever which balances file size against quality. As codec users all we really need to do is pick a pre-baked setting (IMO you only really need low / medium / high) for our application. For JPEG that's usually 70 / 75 / 85, with 90+ reserved for special cases like vector images.

**Comment by jcupitt** [comment]
Oh wait, sorry, what's in these PNGs? Alpha of course won't carry over, I expect you know. Vector art will always look a bit crappy as JPEG, and may not even be smaller. Photographs should work well, but they probably shouldn't be PNGs to start with.

You could consider JXL or WEBP too.

**Comment by shun-recall** [comment]
I'm using C so not Python at the moment.

I'm saving the images at memory buffers and not sure about what libvips functions I need to use to do the SSIM calculation. Could you give me some hints about how to implement it with C?

Thank you very much for all your help John.

**Reply by jcupitt** [reply]
It's just the operations in the "arithmetic" section:

https://www.libvips.org/API/current/libvips-arithmetic.html

If I were you I'd would prototype in python (it's *much* simpler), then once it's working, translate to C. You can run a python program with eg.:

```
$ VIPS_TRACE=1 ./space.py ~/pics/k2.jpg x.v
vips cache+: jpegload filename="/home/john/pics/k2.jpg" out=((VipsImage*) 0x2e4b940) flags=((VipsForeignFlags) VIPS_FOREIGN_SEQUENTIAL) -
vips cache : copy in=((VipsImage*) 0x2e4b940) out=((VipsImage*) 0x2e53cb0) -
vips cache+: sRGB2scRGB in=((VipsImage*) 0x2e4b940) out=((VipsImage*) 0x2e55220) -
vips cache+: cast in=((VipsImage*) 0x2e55220) out=((VipsImage*) 0x2e570c0) format=((VipsBandFormat) VIPS_FORMAT_FLOAT) -
vips cache+: scRGB2XYZ in=((VipsImage*) 0x2e55220) out=((VipsImage*) 0x2e587a0) -
...
```

and it'll show all the low-level libvips operations that are being executed. It can be a bit helpful.

You could consider prototyping in nip2:

https://github.com/libvips/nip2

You can edit equations and watch pixels interactively, but there's a steep learning curve :( I'm making nip4 RN, which ought to be simpler.

**Reply by shun-recall** [reply]
Thank you very much John. I will follow your advice of prototyping with Python.

Something that I'm struggling with is the best way to save an image as JPEG and then to PNG for comparing them with some command line tools like DSSIM (that only supports PNG). Is there any way to avoid saving it to JPG and then load it again and save it as PNG? (I mean a way to avoid the double load-save part)

**Reply by jcupitt** [reply]
You can just save twice, eg.:

```C
VipsImage *image = vips_image_new_from_file("x.jpg", NULL);

vips_image_write_to_file(image, "y.jpg", NULL);
vips_image_write_to_file(image, "y.png", NULL);
```

You can't use streaming (`access="sequential"`) images like this of course.

**Reply by shun-recall** [reply]
Thank you very much @jcupitt
